Adds a -Werror debug Autotools test and updates build modes

derobins opened the pull request:
The Autotools build modes didn't look like they were doing the right thing, so I updated that while I was in there.

Commit:
Adds a -Werror debug Autotools test and updates build modes

Commit:
Un-excludes the workflow files

derobins commented:
There are some gcc 12 warnings that I was not seeing on my gcc 11 VM that cause failures with -Werror. I'll submit a separate PR with gcc 12 warning fixes and merge this PR after that one goes in.

byrnHDF commented on .github/workflows/pr-check.yml:
```diff
@@ -81,6 +86,21 @@ jobs:
+          - name: "-Werror Ubuntu Debug Autotools GCC"
+            artifact: "LinuxDBG.tar.xz"
+            os: ubuntu-latest
+            build_type: "debug"
```
Capital "D" - "Debug"

derobins replied:
Not in the Autotools. We need --enable-build-mode=debug.

Commit:
Merge remote-tracking branch 'canonical/develop' into github_werror

Commit:
Strip out -Werror from CFLAGS when running configure tests

This causes spurious test failures when the configuration test
programs raise warnings.

derobins commented:
This PR should NOT get merged elsewhere. Only develop has been made warning-free.

gnuoyd commented:
Re: commit [5b56fb3](https://github.com/HDFGroup/hdf5/pull/1833/commits/5b56fb362aca39b24095d38771ba9503b416cbc5), it looks like you may not be using `--enable-warnings-as-errors`. That will turn on the errors promotion without breaking `configure`.

opoplawski commented on configure.ac:
```diff
@@ -181,6 +183,14 @@ saved_user_JAVAFLAGS="$JAVAFLAGS"
 
+## Strip out -Werror from CFLAGS since that can cause checks to fail when
+## compiling the test program fails due to warnings
+CFLAGS="`echo $CFLAGS | sed -e 's/-Werror//g'`"
```
This breaks compilation on Fedora which uses `-Werror=format-security`
